File: client-examples/javascript-client/spoke_viewer.mjs
#!/usr/bin/env node
/**
 * Spoke viewer for mayara-server.
 *
 * Connects to a running mayara-server, discovers the first radar,
 * and displays sampled spoke data as ASCII art.
 *
 * The protobuf definition is read from the source tree so it stays
 * in sync with what the server writes.
 *
 * Usage:
 *   node spoke_viewer.mjs [--url http://localhost:6502]
 *
 * Requirements (installed automatically by run.sh):
 *   npm install protobufjs ws
 */

import { dirname, resolve } from "path";
import { fileURLToPath } from "url";
import { parseArgs } from "util";

import protobuf from "protobufjs";
import WebSocket from "ws";

const __dirname = dirname(fileURLToPath(import.meta.url));

// ---------------------------------------------------------------------------
// Load protobuf from source tree
// ---------------------------------------------------------------------------

const PROTO_PATH = resolve(
  __dirname,
  "..",
  "..",
  "src",
  "lib",
  "protos",
  "RadarMessage.proto"
);

const root = protobuf.loadSync(PROTO_PATH);
const RadarMessage = root.lookupType("RadarMessage");

// ---------------------------------------------------------------------------
// REST helpers
// ---------------------------------------------------------------------------

async function fetchJson(url) {
  const res = await fetch(url);
  if (!res.ok) throw new Error(`${url}: ${res.status} ${res.statusText}`);
  return res.json();
}

async function discoverRadar(baseUrl) {
  const data = await fetchJson(
    `${baseUrl}/signalk/v2/api/vessels/self/radars`
  );
  const ids = Object.keys(data);
  if (ids.length === 0) {
    console.error(
      "No radars found. Is the server running with --emulator or a real radar?"
    );
    process.exit(1);
  }
  return { id: ids[0], info: data[ids[0]] };
}

async function fetchCapabilities(baseUrl, radarId) {
  return fetchJson(
    `${baseUrl}/signalk/v2/api/vessels/self/radars/${radarId}/capabilities`
  );
}

// ---------------------------------------------------------------------------
// ASCII art helpers
// ---------------------------------------------------------------------------

const SHADE = " ·:+#@"; // 6 levels

function spokeToAscii(data, width = 64) {
  if (!data || data.length === 0) return " ".repeat(width);
  const step = Math.max(1, data.length / width);
  let chars = "";
  for (let i = 0; i < width; i++) {
    const idx = Math.min(Math.floor(i * step), data.length - 1);
    const val = data[idx];
    const level =
      val > 0
        ? Math.min(SHADE.length - 1, Math.floor((val * SHADE.length) / 64))
        : 0;
    chars += SHADE[level];
  }
  return chars.padEnd(width);
}

function fmtAngle(angle, spokesPerRev) {
  return `${((angle * 360) / spokesPerRev).toFixed(1)}°`.padStart(7);
}

function fmtBearing(bearing, spokesPerRev) {
  if (bearing == null) return "  N/A ";
  return `${((bearing * 360) / spokesPerRev).toFixed(1)}°`.padStart(7);
}

function fmtRange(meters) {
  if (meters >= 1852) return `${(meters / 1852).toFixed(1)} nm`;
  return `${meters} m`;
}

function fmtTime(ms) {
  if (ms == null) return "N/A";
  return new Date(Number(ms)).toISOString().slice(11, 19);
}

function fmtPosition(lat, lon) {
  if (lat == null || lon == null) return "N/A";
  const ns = lat >= 0 ? "N" : "S";
  const ew = lon >= 0 ? "E" : "W";
  return `${Math.abs(lat).toFixed(4)}°${ns} ${Math.abs(lon).toFixed(4)}°${ew}`;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  const { values } = parseArgs({
    options: { url: { type: "string", default: "http://localhost:6502" } },
  });
  const baseUrl = values.url;
  const wsBase = baseUrl.replace(/^http/, "ws");

  // Discover radar
  const { id: radarId, info: radarInfo } = await discoverRadar(baseUrl);
  const caps = await fetchCapabilities(baseUrl, radarId);

  const spokesPerRev = caps.spokesPerRevolution || 2048;
  const maxSpokeLen = caps.maxSpokeLength || 1024;
  const sampleCount = 32;
  const sampleInterval = Math.floor(spokesPerRev / sampleCount);

  // Header
  console.log();
  console.log("=".repeat(78));
  console.log(`  Radar:   ${radarInfo.name || radarId}`);
  console.log(`  Brand:   ${radarInfo.brand || "?"}`);
  if (radarInfo.model) console.log(`  Model:   ${radarInfo.model}`);
  console.log(`  ID:      ${radarId}`);
  console.log(`  Spokes/rev:  ${spokesPerRev}`);
  console.log(`  Max spoke:   ${maxSpokeLen} samples`);
  console.log("=".repeat(78));
  console.log();
  console.log("Connecting to spoke stream...");

  // Build WebSocket URL, replacing host if needed
  let spokeUrl = radarInfo.spokeDataUrl;
  const spokeHost = new URL(spokeUrl).origin;
  const targetHost = new URL(wsBase).origin.replace(/^http/, "ws");
  spokeUrl = spokeUrl.replace(spokeHost, targetHost);
  console.log(`  URL: ${spokeUrl}`);
  console.log();

  // Collect one revolution of spokes
  const seen = new Set();
  const sampled = [];
  let currentRange = 0;
  let currentTime = null;
  let currentLat = null;
  let currentLon = null;
  let revCount = 0;
  let prevAngle = null;

  const ws = new WebSocket(spokeUrl);
  ws.binaryType = "arraybuffer";

  for await (const [data] of on(ws, "message")) {
    if (typeof data === "string") continue;

    const buf = Buffer.isBuffer(data) ? data : Buffer.from(data);
    const msg = RadarMessage.decode(buf);

    let done = false;
    for (const spoke of msg.spokes) {
      // Detect revolution wrap
      if (prevAngle !== null && spoke.angle < prevAngle) {
        revCount++;
        if (revCount >= 1) {
          done = true;
          break;
        }
      }
      prevAngle = spoke.angle;

      currentRange = spoke.range;
      if (spoke.time != null) currentTime = spoke.time;
      if (spoke.lat != null) currentLat = spoke.lat;
      if (spoke.lon != null) currentLon = spoke.lon;

      // Sample this spoke?
      const bucket = Math.floor(spoke.angle / sampleInterval);
      if (!seen.has(bucket)) {
        seen.add(bucket);
        sampled.push(spoke);
      }
    }

    if (done) break;
  }

  ws.terminate();

  // Sort by angle
  sampled.sort((a, b) => a.angle - b.angle);

  // Print range info
  console.log(`  Range:    ${fmtRange(currentRange)} (${currentRange} m)`);
  console.log(`  Time:     ${fmtTime(currentTime)} UTC`);
  console.log(`  Position: ${fmtPosition(currentLat, currentLon)}`);
  console.log();

  // Print spoke field documentation
  console.log("Spoke fields (from RadarMessage.proto):");
  console.log(
    `  angle   - Rotation from bow [0..${spokesPerRev}) clockwise`
  );
  console.log("  bearing - True bearing from North (optional)");
  console.log("  range   - Range of last pixel in meters");
  console.log("  time    - Epoch milliseconds (optional)");
  console.log("  lat/lon - Radar position (optional)");
  console.log("  data    - Pixel intensity bytes (0 = no return)");
  console.log();

  // Legend
  console.log(
    `  ASCII legend:  ${[...SHADE].map((c, i) => `${c}=${i}`).join("  ")}`
  );
  console.log(
    `                 (pixel values are mapped to ${SHADE.length} levels)`
  );
  console.log();

  // Column headers
  console.log(
    `  ${"angle".padStart(7)} ${"brng".padStart(6)} ${"range".padStart(7)} ${"len".padStart(4)}  spoke data`
  );
  console.log(
    `  ${"─".repeat(7)} ${"─".repeat(6)} ${"─".repeat(7)} ${"─".repeat(4)}  ${"─".repeat(64)}`
  );

  for (const spoke of sampled) {
    const angleStr = fmtAngle(spoke.angle, spokesPerRev);
    const bearing =
      spoke.bearing != null ? spoke.bearing : null;
    const bearingStr = fmtBearing(bearing, spokesPerRev);
    const rangeStr = fmtRange(spoke.range).padStart(7);
    const dataLen = String(spoke.data.length).padStart(4);
    const ascii = spokeToAscii(spoke.data);

    console.log(
      `  ${angleStr} ${bearingStr} ${rangeStr} ${dataLen}  ${ascii}`
    );
  }

  console.log();
  console.log(
    `  Sampled ${sampled.length} of ${spokesPerRev} spokes (1 per ${sampleInterval})`
  );
  console.log();
}

/**
 * Simple async iterator for Node.js EventEmitter events.
 * Yields arrays of event arguments until the emitter errors or closes.
 */
function on(emitter, event) {
  const queue = [];
  let resolve;
  let done = false;

  emitter.on(event, (...args) => {
    if (resolve) {
      const r = resolve;
      resolve = null;
      r({ value: args, done: false });
    } else {
      queue.push(args);
    }
  });
  emitter.on("close", () => {
    done = true;
    if (resolve) resolve({ value: undefined, done: true });
  });
  emitter.on("error", (err) => {
    done = true;
    if (resolve) resolve({ value: undefined, done: true });
  });

  return {
    [Symbol.asyncIterator]() {
      return {
        next() {
          if (queue.length > 0) {
            return Promise.resolve({ value: queue.shift(), done: false });
          }
          if (done) return Promise.resolve({ value: undefined, done: true });
          return new Promise((r) => {
            resolve = r;
          });
        },
      };
    },
  };
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});

File: client-examples/python-client/spoke_viewer.py
#!/usr/bin/env python3
"""
Spoke viewer for mayara-server.

Connects to a running mayara-server, discovers the first radar,
and displays sampled spoke data as ASCII art.

Usage:
    python3 spoke_viewer.py [--url http://localhost:6502]

Requirements:
    pip install websockets requests protobuf
"""

import argparse
import asyncio
import math
import struct
import sys
import time
from pathlib import Path

import requests
import ssl
import websockets

# ---------------------------------------------------------------------------
# Load the RadarMessage protobuf definition from the source tree
# ---------------------------------------------------------------------------

def compile_proto():
    """Compile RadarMessage.proto from the source tree and import it."""
    proto_path = Path(__file__).resolve().parent.parent.parent / "src" / "lib" / "protos"
    proto_file = proto_path / "RadarMessage.proto"
    if not proto_file.exists():
        sys.exit(f"Cannot find {proto_file}")

    # Compile into a temp directory so we don't pollute the source tree
    import tempfile
    out_dir = Path(tempfile.mkdtemp(prefix="mayara_proto_"))

    from google.protobuf.compiler import plugin_pb2  # noqa: F401 – just check protobuf is installed
    raise ImportError  # fall through to grpc_tools / shell compile


def load_proto():
    """
    Import the compiled protobuf module.
    Tries grpc_tools first, then shelling out to protoc.
    """
    proto_path = Path(__file__).resolve().parent.parent.parent / "src" / "lib" / "protos"
    proto_file = proto_path / "RadarMessage.proto"
    if not proto_file.exists():
        sys.exit(f"Cannot find {proto_file}")

    import tempfile, importlib
    out_dir = Path(tempfile.mkdtemp(prefix="mayara_proto_"))

    # Try grpc_tools
    try:
        from grpc_tools import protoc as grpc_protoc
        rc = grpc_protoc.main([
            "grpc_tools.protoc",
            f"--proto_path={proto_path}",
            f"--python_out={out_dir}",
            "RadarMessage.proto",
        ])
        if rc != 0:
            raise RuntimeError(f"grpc_tools.protoc returned {rc}")
    except (ImportError, RuntimeError):
        # Fall back to system protoc
        import subprocess
        try:
            subprocess.check_call([
                "protoc",
                f"--proto_path={proto_path}",
                f"--python_out={out_dir}",
                "RadarMessage.proto",
            ])
        except FileNotFoundError:
            sys.exit(
                "Cannot compile RadarMessage.proto.\n"
                "Install one of:\n"
                "  pip install grpcio-tools\n"
                "  brew install protobuf   (or apt install protobuf-compiler)\n"
            )

    sys.path.insert(0, str(out_dir))
    return importlib.import_module("RadarMessage_pb2")


pb2 = load_proto()

# ---------------------------------------------------------------------------
# REST helpers
# ---------------------------------------------------------------------------

def discover_radar(base_url):
    """Fetch the first radar from the API and return (radar_id, radar_info)."""
    r = requests.get(f"{base_url}/signalk/v2/api/vessels/self/radars", verify=False)
    r.raise_for_status()
    data = r.json()
    radars = data
    if not radars:
        sys.exit("No radars found. Is the server running with --emulator or a real radar?")
    radar_id = next(iter(radars))
    return radar_id, radars[radar_id]


def fetch_capabilities(base_url, radar_id):
    """Fetch radar capabilities."""
    r = requests.get(f"{base_url}/signalk/v2/api/vessels/self/radars/{radar_id}/capabilities", verify=False)
    r.raise_for_status()
    return r.json()

# ---------------------------------------------------------------------------
# ASCII art helpers
# ---------------------------------------------------------------------------

SHADE = " ·:+#@"  # 6 levels: nothing, faint, light, medium, strong, max

def spoke_to_ascii(data, width=64):
    """Convert spoke byte data to an ASCII string of the given width."""
    if not data:
        return " " * width
    step = max(1, len(data) / width)
    chars = []
    for i in range(width):
        idx = int(i * step)
        if idx >= len(data):
            break
        val = data[idx]
        level = min(len(SHADE) - 1, val * len(SHADE) // 64) if val > 0 else 0
        chars.append(SHADE[level])
    return "".join(chars).ljust(width)


def format_angle(angle, spokes_per_rev):
    """Format a spoke angle as degrees."""
    deg = angle * 360.0 / spokes_per_rev
    return f"{deg:6.1f}°"


def format_bearing(bearing, spokes_per_rev):
    """Format a bearing as compass degrees."""
    if bearing is None:
        return "  N/A "
    deg = bearing * 360.0 / spokes_per_rev
    return f"{deg:6.1f}°"


def format_range(meters):
    """Format range in human-readable units."""
    if meters >= 1852:
        return f"{meters / 1852:.1f} nm"
    return f"{meters} m"


def format_time(ms):
    """Format epoch millis as HH:MM:SS."""
    if ms is None:
        return "N/A"
    return time.strftime("%H:%M:%S", time.gmtime(ms / 1000))


def format_position(lat, lon):
    """Format lat/lon."""
    if lat is None or lon is None:
        return "N/A"
    ns = "N" if lat >= 0 else "S"
    ew = "E" if lon >= 0 else "W"
    return f"{abs(lat):.4f}°{ns} {abs(lon):.4f}°{ew}"

# ---------------------------------------------------------------------------
# Main spoke viewer
# ---------------------------------------------------------------------------

async def view_spokes(base_url, ws_url):
    # Discover radar
    radar_id, radar_info = discover_radar(base_url)
    caps = fetch_capabilities(base_url, radar_id)

    spokes_per_rev = caps.get("spokesPerRevolution", 2048)
    max_spoke_len = caps.get("maxSpokeLength", 1024)
    sample_count = 32
    sample_interval = spokes_per_rev // sample_count

    # Print header
    print()
    print("=" * 78)
    print(f"  Radar:   {radar_info.get('name', radar_id)}")
    print(f"  Brand:   {radar_info.get('brand', '?')}")
    if radar_info.get("model"):
        print(f"  Model:   {radar_info['model']}")
    print(f"  ID:      {radar_id}")
    print(f"  Spokes/rev:  {spokes_per_rev}")
    print(f"  Max spoke:   {max_spoke_len} samples")
    print("=" * 78)
    print()
    print("Connecting to spoke stream...")

    spoke_url = radar_info["spokeDataUrl"]
    # Replace host if needed (server may advertise localhost)
    if ws_url:
        from urllib.parse import urlparse
        parsed_base = urlparse(ws_url)
        parsed_spoke = urlparse(spoke_url)
        spoke_url = spoke_url.replace(
            f"{parsed_spoke.scheme}://{parsed_spoke.netloc}",
            f"{parsed_base.scheme}://{parsed_base.netloc}",
        )

    print(f"  URL: {spoke_url}")
    print()

    ssl_ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    ssl_ctx.check_hostname = False
    ssl_ctx.verify_mode = ssl.CERT_NONE

    async with websockets.connect(spoke_url, ssl=ssl_ctx if spoke_url.startswith("wss") else None) as ws:
        # Collect one revolution worth of spokes
        seen = set()
        sampled = []
        current_range = 0
        current_time = None
        current_lat = None
        current_lon = None
        rev_count = 0
        prev_angle = None

        while rev_count < 1:
            msg = await ws.recv()
            if isinstance(msg, str):
                continue

            radar_msg = pb2.RadarMessage()
            radar_msg.ParseFromString(msg)

            for spoke in radar_msg.spokes:
                # Detect revolution wrap
                if prev_angle is not None and spoke.angle < prev_angle:
                    rev_count += 1
                    if rev_count >= 1:
                        break
                prev_angle = spoke.angle

                current_range = spoke.range
                if spoke.HasField("time"):
                    current_time = spoke.time
                if spoke.HasField("lat"):
                    current_lat = spoke.lat
                if spoke.HasField("lon"):
                    current_lon = spoke.lon

                # Sample this spoke?
                bucket = spoke.angle // sample_interval
                if bucket not in seen:
                    seen.add(bucket)
                    sampled.append(spoke)

        # Sort by angle
        sampled.sort(key=lambda s: s.angle)

        # Print range info
        print(f"  Range:    {format_range(current_range)} ({current_range} m)")
        print(f"  Time:     {format_time(current_time)} UTC")
        print(f"  Position: {format_position(current_lat, current_lon)}")
        print()

        # Print spoke field documentation
        print("Spoke fields (from RadarMessage.proto):")
        print("  angle   - Rotation from bow [0..{}) clockwise".format(spokes_per_rev))
        print("  bearing - True bearing from North (optional)")
        print("  range   - Range of last pixel in meters")
        print("  time    - Epoch milliseconds (optional)")
        print("  lat/lon - Radar position (optional)")
        print("  data    - Pixel intensity bytes (0 = no return)")
        print()

        # Legend
        print(f"  ASCII legend:  {'  '.join(f'{c}={i}' for i, c in enumerate(SHADE))}")
        print(f"                 (pixel values are mapped to {len(SHADE)} levels)")
        print()

        # Column headers
        hdr_angle = "angle"
        hdr_bearing = " brng"
        hdr_range = " range"
        hdr_len = "len"
        hdr_data = "spoke data"
        print(f"  {hdr_angle:>7s} {hdr_bearing:>6s} {hdr_range:>7s} {hdr_len:>4s}  {hdr_data}")
        print(f"  {'─' * 7} {'─' * 6} {'─' * 7} {'─' * 4}  {'─' * 64}")

        for spoke in sampled:
            angle_str = format_angle(spoke.angle, spokes_per_rev)
            bearing = spoke.bearing if spoke.HasField("bearing") else None
            bearing_str = format_bearing(bearing, spokes_per_rev)
            range_str = format_range(spoke.range)
            data_len = len(spoke.data)
            ascii_art = spoke_to_ascii(spoke.data)

            print(f"  {angle_str} {bearing_str} {range_str:>7s} {data_len:>4d}  {ascii_art}")

        print()
        print(f"  Sampled {len(sampled)} of {spokes_per_rev} spokes (1 per {sample_interval})")
        print()


def main():
    parser = argparse.ArgumentParser(description="Spoke viewer for mayara-server")
    parser.add_argument("--url", default="http://localhost:6502", help="Server base URL")
    args = parser.parse_args()

    ws_url = args.url.replace("http://", "ws://").replace("https://", "wss://")

    try:
        asyncio.run(view_spokes(args.url, ws_url))
    except KeyboardInterrupt:
        print("\nInterrupted.")
    except requests.ConnectionError:
        sys.exit(f"Cannot connect to {args.url}. Is mayara-server running?")


if __name__ == "__main__":
    main()
